javascript/node/selenium-webdriver/bidi/bidi.js

```diff
@@ -0,0 +1,217 @@
+// Licensed to the Software Freedom Conservancy (SFC) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The SFC licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+const { EventEmitter } = require('node:events')
+const WebSocket = require('ws')
+
+const RESPONSE_TIMEOUT = 1000 * 30
+
+class BIDI extends EventEmitter {
+  id = 0
+  isConnected = false
+  events = []
+  browsingContexts = []
+
+  /**
+   * Create a new websocket connection
+   * @param _webSocketUrl
+   */
+  constructor (_webSocketUrl) {
+    super()
+    this._ws = new WebSocket(_webSocketUrl)
+  }
+
+  /**
+   * open connection to fire an open event
+   * @returns {Promise<resolved>}
+   */
+  connect () {
+    return new Promise((resolve) => this._ws.on('open', () => {
+      this.isConnected = true
+      resolve()
+    }))
+  }
+
+  /**
+   * @returns {WebSocket}
+   */
+  get socket () {
+    return this._ws
+  }
+
+  /**
+   * @returns {boolean|*}
+   */
+  get isConnected () {
+    return this.isConnected
+  }
+
+  /**
+   * Sends a bidi request
+   * @param params
+   * @returns {Promise<unknown>}
+   */
+  send (params) {
+    if (!this.isConnected) {
+      throw new Error('No connection to WebDriver Bidi was established')
+    }
+
+    const id = ++this.id
+
+    this._ws.send(JSON.stringify({ id, ...params }))
+
+    return new Promise((resolve, reject) => {
+      const t = setTimeout(() => {
+        reject(new Error(`Request with id ${id} timed out`))
+        handler.off('message', listener)
+      }, RESPONSE_TIMEOUT)
+
+      const listener = (data) => {
+        try {
+          const payload = JSON.parse(data.toString())
+          if (payload.id === id) {
+            clearTimeout(t)
+            handler.off('message', listener)
+            resolve(payload)
+          }
+        } catch (err) {
+          log.error(`Failed parse message: ${err.message}`)
+        }
+      }
+
+      const handler = this._ws.on('message', listener)
+    })
+  }
+
+  /**
+   * Subscribe to events
+   * @param events
+   * @param browsingContexts
+   * @returns {Promise<void>}
+   */
+  async subscribe (events, browsingContexts) {
```

Comment:
i would also provide some better erroring around subscribe/unsubscribe calls 

```
async subscribe(events, browsingContexts) {
  // move to util
  function toArray(anyArugment) {
    if (anyArugment === undefined) {
      return [];
    }
  
    return Array.isArray(anyArugment) ? Array.from(anyArugment) : [anyArugment];
  }

  const argEvents = toArray(events);
  const argBrowsingContexts = toArray(browsingContexts);

  const params = {
    method: 'session.subscribe',
    params: {},
  };

  if (argEvents.length && argEvents.some(event => typeof event !== 'string')) {
    throw new TypeError('events should be string or string array');
  }

  if (argBrowsingContexts.length && argBrowsingContexts.some(context => typeof context !== 'string')) {
    throw new TypeError('browsingContexts should be string or string array');
  }

  if (argEvents.length) {
    this.events.push(...argEvents);

    params.params.events = this.events;
  }

  if (argBrowsingContexts.length) {
    this.browsingContexts.push(...argBrowsingContexts);

    params.params.contexts = this.browsingContexts;
  }

  await this.send(params);
}
```

Reply:
Implemented the required code snippet ! Thank you @potapovDim !